### DISCUSSION TITLE
Can't redirect user to login on session expired

### DISCUSSION THREAD

**Discussion by MihaiIonutB** [nextauthjs/next-auth#3527] [Help] [answered]
### Question 💬

I got a question regarding the use of required session. I have an SSR application and i want to log out users when their session expires.

To begin with I'm setting the maxAge of the session to 20 seconds.
https://next-auth.js.org/configuration/options#session

After the session expires and is removed from cookies if i try to access any page on the application, the session is still valid letting me to do so and NOT redirecting me to login.

How am i supposed to use the onUnauthenticated handler in this case,

`const { data: session, status } = useSession({
    required: true,
    onUnauthenticated() {
        // Handle not authenticated user
    }
});`

https://next-auth.js.org/getting-started/client#require-session

The desired outcome is to take the user back to the login page after the session has expired. Right now this happens only on manually refreshing the page

I've also tried using signOut() inside the handler but that leads to an infinite loading.

### How to reproduce ☕️

Be in an SSR environment 

Set the maxAge of a session to 20 seconds.

Check if onUnauthenticated is called once the session is erased.

`const { data: session, status } = useSession({
    required: true,
    onUnauthenticated() {
        // Handle not authenticated user
    }
});`

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by richardrails** [comment, marked as answer]
you have to set refetchInterval and {required: true} and after that login page is automatically redirected.
```
<SessionProvider session={pageProps.session} refetchInterval={60}>
...
const { data: session } = useSession({required: true})
```

**Reply by balazsorban44** [reply]
This might not work actually, as simply polling the session will make it never expire, unless the browser puts the tab in idle mode and stops the polling. :eyes:

You could set up a client-only `setInterval` that does not go to the backend to refetch the session, just checks if `session.expires` has passed or not. If it did, you could do a redirect to the login page.

**Reply by jasiene** [reply]
Thanks for the suggestion @balazsorban44.

This is what I'm doing, and it works nicely:

```
useEffect(() => {
  const interval = setInterval(() => {
    if (data && isAfter(new Date(), new Date(data.expires))) {
      signIn();
    }
  }, 1000);
  return () => clearInterval(interval);
}, [data]);
```

`isAfter` comes from [date-fns](https://date-fns.org/v2.28.0/docs/isAfter).

**Reply by lucharlier** [reply]
In my case, Im doing the following : 

`  const { status, data } = useSession();
    const router = useRouter();

  useEffect(() => {
    const interval = setInterval(() => {
      console.log(data?.expires);
      if (data && new Date().getTime() > new Date(data.expires).getTime()) {
        router.replace("/login");
      }
    }, 1000);
    return () => {
      clearInterval(interval);
    };
  }, [data?.expires]);`

The problem is that when the expiry date of the cookie change, the data object is not updated. This cause that sometimes, I redirect the user even when it is not needed. 

Does someone know how to update the data object when the expiry date of the cookie is updated ?

**Comment by danvoyce** [comment]
Is this not handled automatically by next-auth? Things have been working as expected for me, i.e redirecting when the session expired. Now for some reason it's not, and my apis are returning 401s and no auto redirect 🤔

**Reply by bensonjm** [reply]
I'm facing this too. Can anyone help with this?

**Reply by resulselmani** [reply]
`const { data: session } = useSession({
    required: true,
    onUnauthenticated() {
      void router.push('/login')
    }
  })`
  
  this should work 👍

**Reply by giannischiout** [reply]
How can you handle this in a server component in next14 ?

```
import { authOptions } from "@/app/api/auth/[...nextauth]/route";
import AuthComponent from "../_components/AuthComponent";
export default async function Page() {
    const  session = await getServerSession(authOptions);
   
    if(!session) {
        return (
            <AuthComponent redirectTo="/login" />
        )
    }
    return (
        <div>
            <CustomReport />
        </div>
    )
}
```

**Reply by resulselmani** [reply]
I believe it should go something like this: 

```
import { getServerSession } from "next-auth/next";
import { authOptions } from "@/app/api/auth/[...nextauth]/route";
import AuthComponent from "../_components/AuthComponent";
import CustomReport from "../components/CustomReport"; // Assuming you have this component

export default function Page({ session }) {
    if (!session) {
        return <AuthComponent redirectTo="/login" />;
    }

    return (
        <div>
            <CustomReport />
        </div>
    );
}


export async function getServerSideProps(context) {
    const session = await getServerSession(context.req, context.res, authOptions);

    // If there's no session, the page component will handle rendering the AuthComponent
    return {
        props: { session }, // Pass session to the page component as a prop
    };
}
```
